crates/driver/src/boundary/settlement.rs

```diff
@@ -245,6 +248,115 @@ impl Settlement {
             solver: self.solver,
         })
     }
+
+    pub fn with_protocol_fees(&mut self, solution: &competition::Solution) {
+        struct PricedTrade {
+            sell_price: eth::U256,
+            buy_price: eth::U256,
+        }
+
+        let boundary_trades = self
+            .inner
+            .encoder
+            .user_trades()
+            .map(|trade| {
+                (
+                    trade.data.order.metadata.uid.0.into(),
+                    PricedTrade {
+                        sell_price: trade.sell_token_price,
+                        buy_price: trade.buy_token_price,
+                    },
+                )
+            })
+            .collect::<HashMap<Uid, PricedTrade>>();
+
+        let new_boundary_trades = solution
+            .trades()
+            .iter()
+            .filter_map(|trade| {
+                match trade {
+                    solution::Trade::Jit(_) => None,
+                    solution::Trade::Fulfillment(fulfillment) => {
+                        let protocol_fee_factor = 0.5;
+                        let protocol_fee_cap = 0.05;
+
+                        let order = fulfillment.order();
+
+                        // Only apply fees to limit orders.
+                        if !matches!(order.kind, order::Kind::Limit) {
+                            return None;
+                        }
+
+                        let PricedTrade {
+                            sell_price,
+                            buy_price,
+                        } = boundary_trades.get(&order.uid).unwrap().clone();
+                        let sell_amount = *buy_price;
+                        let buy_amount = *sell_price;
+
+                        let executed_amount = match order.side {
+                            order::Side::Sell => fulfillment
+                                .executed()
+                                .0
+                                .checked_add(fulfillment.solver_fee().0)
```

Comment:
are you sure executed doesn't already contain the solver fee?

Reply:
Yes, double checked. Will add a comment to the field.